src/backend/model_deployments/azure.py

```diff
@@ -53,47 +60,109 @@ def is_available(cls) -> bool:
         return all([os.environ.get(var) is not None for var in AZURE_ENV_VARS])
 
     def invoke_chat(self, chat_request: CohereChatRequest, **kwargs: Any) -> Any:
-        yield self.client.chat(
-            **chat_request.model_dump(exclude={"stream"}),
-            **kwargs,
+        start_time = time.perf_counter()
```

Comment:
Same comment for azure/bedrock/cohere/sagemaker, but I think the metrics logic should be refactored out of the actual Model deployment implementation for 2 reasons: DRY and separation of concerns. 

Suggestion:

You can implement some sort of decorator function that takes as arg the Model Deployment chat method (eg, `CoherePlatform.invoke_chat`), and implements the logic of creating the MetricsData, try/except/finally etc. 

Then the idea would be that you could have the metrics in a separate file somewhere (perhaps services?)

And implement your decorator like: 

```
import time
from functools import wraps
from typing import Any, Callable, Dict

def collect_metrics(func: Callable) -> Callable:
    @wraps(func)
    def wrapper(self, chat_request: 'CohereChatRequest', **kwargs: Any) -> Any:
             #.. do metrics stuff
             response = {}
             try:
                      response = func(self, chat_request, **kwargs)
                      response_dict = to_dict(response)
             except..
             finally.. etc
```

Then your methods could just be decorated with this 

```
@collect_metrics
def invoke_chat(etc):
```

Reply:
Would help clean up the code a lot!

Reply:
The `report_metrics` code also looks fine to me, you would probably just move it to the same file `collect_metrics` live, or put them into a class (although I'm not too familiar with encapsulating decorators in a class, the syntax might be different)

Reply:
Great suggestion, the code looks much better now. I kept the decorators specific to each function because they have some particularities and I thought it would be confusing to keep everything together (i.e. streaming and non streaming responses)